backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/tournament/TournamentService.java

```diff
@@ -0,0 +1,343 @@
+
+        return new TournamentDto(tournament);
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<TournamentDto> getAllTournaments(User user) {
+        Set<CourseExecution> set = user.getCourseExecutions();
+        return tournamentRepository.getAllTournaments(set).stream().map(TournamentDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<TournamentDto> getOpenedTournaments(User user) {
+        Set<CourseExecution> set = user.getCourseExecutions();
+        return tournamentRepository.getOpenedTournaments(set).stream().map(TournamentDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<TournamentDto> getClosedTournaments(User user) {
+        Set<CourseExecution> set = user.getCourseExecutions();
+        return tournamentRepository.getClosedTournaments(set).stream().map(TournamentDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<TournamentDto> getUserTournaments(User user) {
+        return tournamentRepository.getUserTournaments(user.getId()).stream().map(TournamentDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public void joinTournament(Integer userId, TournamentDto tournamentDto, String password) {
+        User user = checkUser(userId);
+        Tournament tournament = checkTournament(tournamentDto.getId());
+
+        if (DateHandler.now().isAfter(tournament.getEndTime())) {
+            throw new TutorException(TOURNAMENT_NOT_OPEN, tournament.getId());
+        }
+
+        if (tournament.getState() == Tournament.Status.CANCELED) {
+            throw new TutorException(TOURNAMENT_CANCELED, tournament.getId());
+        }
+
+        if (tournament.getParticipants().contains(user)) {
+            throw new TutorException(DUPLICATE_TOURNAMENT_PARTICIPANT, user.getUsername());
+        }
+
+        if (!user.getCourseExecutions().contains(tournament.getCourseExecution())) {
+            throw new TutorException(STUDENT_NO_COURSE_EXECUTION, user.getId());
+        }
+
+        if (tournament.isPrivateTournament() && !password.equals(tournament.getPassword())) {
+            throw new TutorException(WRONG_TOURNAMENT_PASSWORD, tournament.getId());
+        }
+
+        tournament.addParticipant(user);
+
+        if (!tournament.hasQuiz()) {
```

Comment:
only create quiz when it starts... and write this code in a different method because it will become a strategy in the future

Reply:
Solved